### PR TITLE
Run results postgres controls (create / send / delete)

### PR DESCRIPTION
**Pull request by neomatamune**
- Create DB on Run Start [PROD-13251]
- Delete DB when Run is deleted [PROD-13253]
- Create table and send data on /results/send calls [PROD-13246]
  - When data is empty fails
  - When data has new columns alters table to take new columns into account
  - Separate data into first level columns with type inference
    - boolean -> BOOLEAN
    - int / float -> NUMERICAL
    - string -> TEXT
    - Anything else -> JSONB
  - When column has incompatible types during the first call store column as JSONB data
  - When column has incompatible types on following calls fails
- Update rights for each system user as required 
  - reader : `CONNECT` on DB + `SELECT` on Tables
  - writer : `CONNECT` on DB + `CREATE` on Schema `public`

[PROD-13251]: https://cosmo-tech.atlassian.net/browse/PROD-13251?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ
[PROD-13253]: https://cosmo-tech.atlassian.net/browse/PROD-13253?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ
[PROD-13246]: https://cosmo-tech.atlassian.net/browse/PROD-13246?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

**Review by sjoubert — CHANGES_REQUESTED** [review]

File: run/src/main/kotlin/com/cosmotech/run/exceptions/RunExceptionHandling.kt [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+class RunExceptionHandling {
+
+  private val httpStatusCodeTypePrefix = "https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/"
+
+  @ExceptionHandler(SQLException::class)
+  fun handleJedisDataException(exception: SQLException): ProblemDetail {
+    var response = ProblemDetail.forStatus(HttpStatus.BAD_REQUEST)
+    val badRequestError = HttpStatus.BAD_REQUEST
+    response.type = URI.create(httpStatusCodeTypePrefix + badRequestError.value())
+    response.title = "Bad SQL Request"
```
[the remark this answers is not in the document]
Comment (neomatamune) [reply run/src/main/kotlin/com/cosmotech/run/exceptions/RunExceptionHandling.kt:25]: Just some copy paste I forgot to rename, will change it

File: run/src/main/kotlin/com/cosmotech/run/service/RunServiceImpl.kt [hunk 2]
```diff
@@ -91,35 +130,151 @@ class RunServiceImpl(
...
+            .prepareStatement(
+                "CREATE TABLE \"$dataTableName\" " +
+                    "( ${dataKeys.joinToString(separator = ", ") { "\"$it\" ${dataKeyType[it]!!}" }} ) ")
+            .executeUpdate()
+
+        logger.debug("Creating new table $dataTableName for run $runId")
+
+        // Grant select rights to the reader account
+        connection
+            .prepareStatement("GRANT SELECT ON TABLE \"$dataTableName\" TO $readerStorageUsername ")
```
[the remark this answers is not in the document]
Comment (neomatamune) [reply run/src/main/kotlin/com/cosmotech/run/service/RunServiceImpl.kt:186]: It was in an effort to try to reduce at maximum the rights given to the reader

File: run/src/main/kotlin/com/cosmotech/run/service/RunServiceImpl.kt [hunk 3]
```diff
@@ -91,35 +130,151 @@ class RunServiceImpl(
...
 
-    if (!runDBJdbcTemplate.existTable(tableName)) {
-      runDBJdbcTemplate.createCustomDataTable(tableName)
+      logger.debug("Inserted ${data.size} rows in table $dataTableName for run $runId")
+      connection.commit()
+      connection.close()
+    } catch (e: SQLException) {
+      connection.rollback()
+      connection.close()
+      throw e
```
[the remark this answers is not in the document]
Comment (neomatamune) [reply run/src/main/kotlin/com/cosmotech/run/service/RunServiceImpl.kt:258]: can a `finally` be called even after a `throw` ?

File: run/src/main/kotlin/com/cosmotech/run/service/RunServiceImpl.kt [hunk 4]
```diff
@@ -250,6 +406,28 @@ class RunServiceImpl(
...
+        .executeUpdate()
+    connection
+        .prepareStatement("GRANT CONNECT ON DATABASE \"$runId\" TO $writerStorageUsername")
+        .executeUpdate()
+
+    connection
+        .prepareStatement("GRANT CREATE ON SCHEMA public to $writerStorageUsername")
+        .executeUpdate()
+    connection.commit()
+    connection.close()
```
[the remark this answers is not in the document]
Comment (neomatamune) [reply run/src/main/kotlin/com/cosmotech/run/service/RunServiceImpl.kt:430]: same as above : can a `finally` be called even after a `throw` ?

File: run/src/main/kotlin/com/cosmotech/run/config/RunStorageConfig.kt [hunk 5]
```diff
@@ -45,6 +50,21 @@ class RunStorageConfig {
...
+  @Bean
+  fun writerRunStorageDatasource(): DriverManagerDataSource {
+    val dataSource =
+        DriverManagerDataSource(storageHost, writerStorageUsername, writerStoragePassword)
+    dataSource.setDriverClassName(jdbcdriverClass)
+    return dataSource
+  }
+
+  @Bean
+  fun writerRunStorageTemplate(
```
[the remark this answers is not in the document]
Comment (neomatamune) [reply run/src/main/kotlin/com/cosmotech/run/config/RunStorageConfig.kt:62]: no idea, copied the `@Bean` from other definition you made, never looked at what it does, so tell me, what does it do ? is it important or I could just remove it and never think about it again ?

[the remark this answers is not in the document]
Comment (jreynard-code) [reply run/src/main/kotlin/com/cosmotech/run/config/RunStorageConfig.kt:62]: Both can be removed as they are not used. Only adminuser should be kept I think. As we create a jdbc connection at runtime directly with username/password

File: run/src/main/kotlin/com/cosmotech/run/service/RunServiceImpl.kt (hunk 2 above)
[the remark this answers is not in the document]
Comment (sjoubert) [reply run/src/main/kotlin/com/cosmotech/run/service/RunServiceImpl.kt:186]: Ok, I have no real points to do that another way.

File: run/src/main/kotlin/com/cosmotech/run/service/RunServiceImpl.kt (hunk 3 above)
[the remark this answers is not in the document]
Comment (jreynard-code) [reply run/src/main/kotlin/com/cosmotech/run/service/RunServiceImpl.kt:258]: Yep.
Here is a little snippet:

```
fun main() {
    println("Start")

    try {
        throwException()
    } catch (e:IllegalStateException) {
        println("catch the error")
        throw e
    } finally {
        println("finally")
    }
}

fun throwException() {
    throw IllegalStateException("error")
}
```

Output is :
```
Start
catch the error
finally
```

File: run/src/main/kotlin/com/cosmotech/run/service/RunServiceImpl.kt (hunk 4 above)
[the remark this answers is not in the document]
Comment (jreynard-code) [reply run/src/main/kotlin/com/cosmotech/run/service/RunServiceImpl.kt:430]: yep same remark as above

**Review by sjoubert — APPROVED** [review]

**Pull request merged**